src/compiler/xslt/invoke-compiled/group-invocation.xsl

```diff
@@ -0,0 +1,76 @@
+                xmlns:x="http://www.jenitennison.com/xslt/xspec"
+                xmlns:xs="http://www.w3.org/2001/XMLSchema"
+                xmlns:xsl="http://www.w3.org/1999/XSL/Transform"
+                exclude-result-prefixes="#all"
+                version="3.0">
+
+   <!--
+      mode="x:group-invocation"
+      This mode groups the invocation instructions of compiled x:scenario into a single xsl:for-each.
+   -->
+
+   <xsl:template match="document-node()" as="element()*" mode="x:group-invocation">
+      <!-- xsl:call-template elements originating from x:scenario -->
+      <xsl:variable name="scenario-invokers" as="element(xsl:call-template)*"
+         select="xsl:call-template[processing-instruction(origin) eq 'scenario']" />
+
+      <!-- All the invocation instructions originating from x:scenario must be adjacent so that they
+         can be grouped in a single group. -->
+      <xsl:for-each select="
+            node()
+            [not(. intersect $scenario-invokers)]
+            [. >> $scenario-invokers[1]]
+            [$scenario-invokers[last()] >> .]">
+         <xsl:message terminate="yes">
+            <xsl:call-template name="x:prefix-diag-message">
+               <xsl:with-param name="message"
+                  select="'Unexpected node between invocation instructions of compiled scenario.'" />
+            </xsl:call-template>
+         </xsl:message>
+      </xsl:for-each>
+
+      <!-- Group the adjacent xsl:call-template elements originating from x:scenario. The other
+         nodes (actually, xsl:call-template elements (from x:expect) and x:variable elements (from
+         x:param and x:variable)) are grouped individually. -->
+      <xsl:for-each-group select="node() treat as element()*" group-adjacent="
+            if (. intersect $scenario-invokers) then
+               0
+            else
+               position()">
+         <xsl:apply-templates select="." mode="#current" />
+      </xsl:for-each-group>
+   </xsl:template>
+
+   <xsl:template match="xsl:call-template[processing-instruction(origin) eq 'scenario']"
+      as="element()+" mode="x:group-invocation">
+      <!-- x:description or x:scenario invoking the current xsl:call-template -->
+      <xsl:param name="tunnel_invoker-description-or-scenario" as="element()" required="yes"
+         tunnel="yes" />
+
+      <xsl:variable name="child-scenario-count" as="xs:integer" select="current-group() => count()" />
+
+      <for-each select="1 to {$child-scenario-count}">
+         <!-- TODO: multi-threading
+         <xsl:if test="$user-wants-to-enable-threads">
+            <xsl:attribute name="saxon:threads" namespace="{$x:saxon-namespace}"
+               select="'{$' || $thread-count-variable-uqname || '}'" />
+         </xsl:if>
+         -->
+         <choose>
+            <xsl:for-each select="current-group()">
+               <when test=". eq {count(current-group()[current() >> .] | .)}">
```

Comment:
For me, `1+count(preceding-sibling::*)` is easier to understand than your expression in curly braces. Are they equivalent?

Reply:
`1+count(preceding-sibling::*)` would count irrelevant elements and compile https://github.com/xspec/xspec/blob/10128c234f1dea22a4df0e706aa05ff203b0be7a/test/variable.xspec#L169-L181 into 
```xslt
         <xsl:for-each select="1 to 1">
            <xsl:choose>
               <xsl:when test=". eq 3">
                  <xsl:call-template name="x:scenario4-scenario2-scenario1">
                     <xsl:with-param name="myv:outer"
                                     select="$myv:outer"/>
                  </xsl:call-template>
               </xsl:when>
            </xsl:choose>
         </xsl:for-each>
```

I added d26fb889f24285fb14c92bcfb2ce2a9a5eacd863 for safeguard against this kind of possible errors.

d2d34842899ad8ace830ce1fc276054c706f0fef simplifies the expression using `position()`. I don't know why I didn't write it in the first place.

Reply:
I was just going to say never mind because my expression was not equivalent to yours. You got here first. :)